File: spsc_queue.go
package spscqueue

import (
	"errors"
	"sync/atomic"
	"syscall"
	"unsafe"

	"gopkg.in/ro-ag/posix.v1"
)

const (
	cacheLineSize = 64       // cache line 的大小
	shmHeaderLen  = 1024 * 4 // 共享内存的头长度，是一个 page 的长度
	minQueueBytes = 1024     // 队列最小的字节数
	itemHeaderLen = 4        // 每个 item 使用 4 字节表示长度
	minLeftLen    = 7        // 尾部剩余的最小字节数
	itemLenMask   = 3        // item的长度的掩码
	shmMode       = 0666     // 共享内存的权限
)

// SpscQueue 共享内存队列的结构
// 共计 1024*4 字节，放置于共享内存的头部
type SpscQueue struct {
	consumer  uint64
	Reserved1 [cacheLineSize - 8]byte // 防止伪共享
	producer  uint64
	Reserved2 [cacheLineSize - 8]byte
	mask      uint64 // 队列长度必须是 2 的 n 次幂
	Reserved3 [shmHeaderLen - cacheLineSize*2 - 8]byte
	// todo: 增加 eventfd
	// todo: 文件锁
}

// NewSpscQueueFromShm 构造队列
// @param shmName 共享内存名称
// @param queueBytes 队列的字节数，必须是 2 的 n 次幂
// @param isCreate 共享内存不存在的时候，是否创建
func NewSpscQueueFromShm(shmName string, queueBytes uint64,
	isCreate bool) (q *SpscQueue, err error) {
	if RoundPowerOfTwo(queueBytes) != queueBytes {
		err = ErrOfBadQueueSize
		return
	}
	var fd int
	var data []byte
	var isFirstTime bool
	fd, err = posix.ShmOpen(shmName, syscall.O_RDWR, shmMode)
	if err != nil {
		if !errors.Is(err, syscall.ENOENT) {
			err = ErrOfOpenShmFail
			return
		}
		if !isCreate {
			err = ErrOfShmNotExist
			return
		}
		fd, err = posix.ShmOpen(shmName, posix.O_CREAT|posix.O_RDWR, shmMode)
		if err != nil {
			err = errors.Join(ErrOfCreateShmFail, err)
			return
		}
		err = posix.Ftruncate(fd, int(queueBytes+shmHeaderLen))
		if err != nil {
			_ = posix.Close(fd)
			err = errors.Join(ErrOfTruncateShm, err)
			return
		}
		isFirstTime = true
	}
	// map
	data, _, err = posix.Mmap(nil, int(queueBytes+shmHeaderLen),
		posix.PROT_READ|posix.PROT_WRITE, posix.MAP_SHARED, fd, 0)
	if err != nil {
		_ = posix.Close(fd)
		err = errors.Join(ErrOfMMapError, err)
		return
	}
	_ = posix.Close(fd)
	//
	q = (*SpscQueue)(unsafe.Pointer(&data[0]))
	if isFirstTime {
		q.consumer = 0
		q.producer = 0
		q.mask = queueBytes - 1
	} else {
		if q.mask+1 != queueBytes {
			err = ErrOfBadMaskOfOpenedShm
			_ = posix.Munmap(data)
			return
		}
	}
	return
}

// NewSpscQueue 在堆上分配 SpscQueue
func NewSpscQueue(queueBytes uint64) (*SpscQueue, error) {
	if RoundPowerOfTwo(queueBytes) != queueBytes {
		return nil, ErrOfBadQueueSize
	}
	totalBytes := shmHeaderLen + queueBytes + minLeftLen
	buf := make([]byte, totalBytes)
	padding := uint64(uintptr(unsafe.Pointer(&buf[0]))) & minLeftLen
	q := (*SpscQueue)(unsafe.Add(unsafe.Pointer(&buf[0]), padding))
	q.consumer = 0
	q.producer = 0
	q.mask = queueBytes - 1
	return q, nil
}

// Alloc 从队尾分配内存
// @param bytes 一个块的大小。一个块不超过 4gb
func (q *SpscQueue) Alloc(needBytes uint32) (out []byte, newTail uint64, err error) {
	// 从队尾申请内存
	if uint64(needBytes) > q.mask/2 {
		err = ErrOfBytesTooLarge
		return
	}
	if needBytes == 0 {
		err = ErrOfBadParam
		return
	}
	for {
		curTail := atomic.LoadUint64(&q.producer)
		curHead := atomic.LoadUint64(&q.consumer) // head 读到版本 2 会怎么样? 会更加准确
		if (curTail+1)&q.mask == curHead {        // ??? 有必要吗
			err = ErrOfNotEnoughSpace // queue is full
			return
		}
		padding := curTail & itemLenMask // 按照 4 字节对齐
		if curTail >= curHead {
			if curTail+minLeftLen+uint64(needBytes) >= q.Len() {
				if q.consumer == 0 {
					err = ErrOfNotEnoughSpace
					return
				}
				if !atomic.CompareAndSwapUint64(&q.producer, curTail, 0) {
					continue
				}
				if curTail+padding+itemHeaderLen <= q.mask {
					// 尾部空间不够的时候，绕到头部
					nextItemPtr := (*uint32)(q.GetPointer(curTail + padding))
					atomic.StoreUint32(nextItemPtr, 0)
				}
				continue
			}
		} else {
			if curTail+minLeftLen+uint64(needBytes)+1 > curHead {
				err = ErrOfNotEnoughSpace
				return
			}
		}
		newTail = curTail + padding + itemHeaderLen + uint64(needBytes)
		nextItemPtr := (*uint32)(q.GetPointer(curTail + padding))
		oldValue := atomic.LoadUint32(nextItemPtr)
		if !atomic.CompareAndSwapUint32(nextItemPtr, oldValue, needBytes) { // todo: 理论上用 atomic_store 就够了
			continue
		}
		out = unsafe.Slice((*byte)(q.GetPointer(curTail+padding+itemHeaderLen)), needBytes)
		return
	}
}

// Len 数据区域的字节数
func (q *SpscQueue) Len() uint64 {
	return q.mask + 1
}

// CommitProduce 提交 Alloc, 尾部位置移动到新位置
func (q *SpscQueue) CommitProduce(newTail uint64) bool {
	curTail := atomic.LoadUint64(&q.producer)
	return atomic.CompareAndSwapUint64(&q.producer, curTail, newTail)
}

// Produce 把 Alloc + CommitProduce 合并到一个方法
func (q *SpscQueue) Produce(src []byte) error {
	dst, newTail, err := q.Alloc(uint32(len(src)))
	if err != nil {
		return err
	}
	copy(dst, src)
	if !q.CommitProduce(newTail) {
		return ErrOfCommitProduceFail
	}
	return nil
}

// GetPointer 返回 queue 的数据区的指针
func (q *SpscQueue) GetPointer(offset uint64) unsafe.Pointer {
	if offset > q.mask {
		panic("offset out of range")
	}
	return unsafe.Add(unsafe.Pointer(q), shmHeaderLen+offset)
}

// IsFull 是否队满
func (q *SpscQueue) IsFull() bool {
	return (atomic.LoadUint64(&q.producer)+1)&q.mask == atomic.LoadUint64(&q.consumer)
}

// IsEmpty 是否队空
func (q *SpscQueue) IsEmpty() bool {
	return atomic.LoadUint64(&q.consumer) == atomic.LoadUint64(&q.producer)
}

// GetOne 读队列头部的 item
func (q *SpscQueue) GetOne() (data []byte, newHead uint64, err error) {
	for {
		curHead := atomic.LoadUint64(&q.consumer)
		curTail := atomic.LoadUint64(&q.producer) // 读到版本 2 无影响
		if curHead == curTail {
			err = ErrOfQueueIsEmpty
			return
		}
		if curHead+minLeftLen >= q.Len() { // 最右侧的 7 字节不会使用
			if curTail > curHead {
				// todo: 一般不可能出现这种情况
				atomic.CompareAndSwapUint64(&q.consumer, curHead, curTail)
			} else {
				atomic.CompareAndSwapUint64(&q.consumer, curHead, 0)
			}
			continue
		}
		padding := curHead & itemLenMask // 按照 4 字节对齐
		nextItemPtr := (*uint32)(q.GetPointer(curHead + padding))
		itemLen := atomic.LoadUint32(nextItemPtr)
		if itemLen == 0 {
			// 0 是特殊标记，表示从这里到尾部都没有数据
			if curTail > curHead {
				//todo: 单元测试中构造这种临界点
				atomic.CompareAndSwapUint64(&q.consumer, curHead, curTail)
			} else {
				atomic.CompareAndSwapUint64(&q.consumer, curHead, 0)
			}
			continue
		}
		newHead = curHead + padding + itemHeaderLen + uint64(itemLen)
		if curTail < curHead {
			if curHead+minLeftLen+uint64(itemLen) > q.Len() {
				panic("impossible error: out of range")
			}
		} else {
			if newHead > curTail {
				panic("impossible error: out of range")
			}
		}
		data = unsafe.Slice((*byte)(q.GetPointer(curHead+padding+itemHeaderLen)), itemLen)
		return
	}
}

// CommitConsume 提交 GetOne()，把 head 修改到新位置
func (q *SpscQueue) CommitConsume(newHead uint64) bool {
	curHead := atomic.LoadUint64(&q.consumer)
	return atomic.CompareAndSwapUint64(&q.consumer, curHead, newHead)
}

// Consume 整合 GetOne() + CommitConsume()
func (q *SpscQueue) Consume(dst []byte) (out []byte, err error) {
	var src []byte
	var newHead uint64
	src, newHead, err = q.GetOne()
	if err != nil {
		return
	}
	out = append(dst, src...)
	if !q.CommitConsume(newHead) {
		err = ErrOfCommitConsumeFail
	}
	return
}

// Unmap 取消共享内存映射
func (q *SpscQueue) Unmap() error {
	data := unsafe.Slice((*byte)(unsafe.Pointer(q)), q.Len()+shmHeaderLen)
	return posix.Munmap(data)
}

File: c/src/spsc_queue.c
#include "spsc_queue.h"

#define itemHeaderLen 4
#define minLeftLen 7
#define itemLenMask 3

#define ATOMIC_LOAD(ptr) __atomic_load_n(ptr, __ATOMIC_SEQ_CST)
#define ATOMIC_CAS(ptr, expectedPtr, desired)                                  \
  __atomic_compare_exchange_n(ptr, expectedPtr, desired, 0, __ATOMIC_SEQ_CST,  \
                              __ATOMIC_SEQ_CST)

bool IsPowerOfTwo(uint64_t n) { return n > 0 && (n & (n - 1)) == 0; }

uint64_t RoundPowerOfTwo(uint64_t n) {
  if (n < minQueueBytes) {
    return minQueueBytes;
  }
  if (IsPowerOfTwo(n)) {
    return n;
  }
  int leading_zeros = __builtin_clz(n);
  return 1 << (64 - leading_zeros);
}

int NewSpscQueueFromShm(const char *shmName, uint64_t queueSize, bool isCreate,
                        int *outErrorNumber, SpscQueue **outQueue) {
  if (!IsPowerOfTwo(queueSize)) {
    return ErrOfBadParam;
  }
  int shm_fd = shm_open(shmName, O_RDWR, 0666);
  bool isFirstTime = false;
  if (shm_fd == -1) {
    *outErrorNumber = errno;
    if (*outErrorNumber != ENOENT) {
      return ErrOfShmOpen;
    }
    // 考虑是否创建
    if (!isCreate) {
      return ErrOfShmNotExists;
    }
    isFirstTime = true;
    shm_fd = shm_open(shmName, O_CREAT | O_RDWR, 0666);
    if (shm_fd == -1) {
      *outErrorNumber = errno;
      return ErrOfShmCreateFail;
    }
    int ret = ftruncate(shm_fd, queueSize + shmHeaderLen);
    if (ret == -1) {
      *outErrorNumber = errno;
      close(shm_fd);
      return ErrOfShmTuncate;
    }
  }
  // 将共享内存映射到进程地址空间
  void *shm_ptr = mmap(0, queueSize + shmHeaderLen, PROT_READ | PROT_WRITE,
                       MAP_SHARED, shm_fd, 0);
  if (shm_ptr == MAP_FAILED) {
    *outErrorNumber = errno;
    close(shm_fd);
    return ErrOfMMap;
  }
  close(shm_fd);
  shm_fd = 0;
  SpscQueue *q = (SpscQueue *)shm_ptr;
  if (isFirstTime) {
    q->Producer = 0;
    q->Consumer = 0;
    q->Mask = queueSize - 1;
  }
  *outQueue = q;
  return NewQueueSuccess;
}

uint64_t Len(const SpscQueue *q) { return q->Mask + 1; }

bool IsFull(const SpscQueue *q) {
  uint64_t curHead = ATOMIC_LOAD(&(q->Consumer));
  uint64_t curTail = ATOMIC_LOAD(&(q->Producer));
  return ((curTail + 1) & q->Mask) == curHead;
}

bool IsEmpty(const SpscQueue *q) {
  uint64_t curHead = ATOMIC_LOAD(&(q->Consumer));
  uint64_t curTail = ATOMIC_LOAD(&(q->Producer));
  return curHead == curTail;
}

int Alloc(SpscQueue *q, uint32_t bytes, void **outPtr, uint64_t *outNewTail) {
  uint64_t needBytes = (uint64_t)bytes;
  if (needBytes > q->Mask / 2) {
    return ErrOfBytesTooLarge;
  }
  if (bytes == 0) {
    return ErrOfBadParamForAlloc;
  }
  for (;;) {
    uint64_t curHead = ATOMIC_LOAD(&(q->Consumer));
    uint64_t curTail = ATOMIC_LOAD(&(q->Producer));
    uint64_t padding = curTail & itemLenMask;
    if (curTail >= curHead) {
      if (curTail + minLeftLen > q->Mask) {
        if (curHead == 0) {
          return ErrOfNotEnoughSpace;
        }
        ATOMIC_CAS(&(q->Producer), &curTail, 0);
        continue;
      }
      if (curTail + minLeftLen + needBytes > q->Mask) {
        if (curHead == 0) {
          return ErrOfNotEnoughSpace;
        }
        // 尾部空间不够的时候，绕到头部
        uint32_t *nextItemPtr =
            (uint32_t *)((uint8_t *)q + shmHeaderLen + curTail + padding);
        uint32_t oldValue = ATOMIC_LOAD(nextItemPtr);
        if (!ATOMIC_CAS(nextItemPtr, &oldValue,
                        0)) { // todo: atomic store 也可以
          continue;
        }
        if (!ATOMIC_CAS(&(q->Producer), &curTail, 0)) {
          uint32_t rollback = 0;
          ATOMIC_CAS(nextItemPtr, &rollback, oldValue);
        }
        continue;
      }
    } else {
      if (curTail + minLeftLen + needBytes + 1 > curHead) {
        return ErrOfNotEnoughSpace;
      }
    }
    *outNewTail = curTail + padding + itemHeaderLen + needBytes;
    uint32_t *nextItemPtr =
        (uint32_t *)((uint8_t *)q + shmHeaderLen + curTail + padding);
    uint32_t oldValue = ATOMIC_LOAD(nextItemPtr);
    if (!ATOMIC_CAS(nextItemPtr, &oldValue, needBytes)) {
      continue;
    }
    *outPtr = (uint8_t *)q + shmHeaderLen + curTail + padding + itemHeaderLen;
    return AllocSuccess;
  }
}

bool CommitProduce(SpscQueue *q, uint64_t newTail) {
  uint64_t curTail = ATOMIC_LOAD(&(q->Producer));
  return ATOMIC_CAS(&(q->Producer), &curTail, newTail);
}

int Produce(SpscQueue *q, const void *src, uint32_t len) {
  void *dst = NULL;
  uint64_t newTail = 0;
  int ret = Alloc(q, len, &dst, &newTail);
  if (ret != AllocSuccess) {
    return ret;
  }
  memcpy(dst, src, len);
  if (!CommitProduce(q, newTail)) {
    return ErrOfCommitFail;
  }
  return AllocSuccess;
}

int GetOne(SpscQueue *q, void **dst, uint32_t *outLen, uint64_t *newHead) {
  for (;;) {
    uint64_t curHead = ATOMIC_LOAD(&(q->Consumer));
    uint64_t curTail = ATOMIC_LOAD(&(q->Producer));
    if (curHead == curTail) {
      return ErrOfQueueIsEmpty;
    }
    if (curHead + minLeftLen > q->Mask) {
      if (curTail > curHead) {
        ATOMIC_CAS(&(q->Consumer), &curHead, curTail);
      } else {
        ATOMIC_CAS(&(q->Consumer), &curHead, 0);
      }
      continue;
    }
    uint64_t padding = curHead & itemLenMask;
    uint32_t *nextItemPtr =
        (uint32_t *)((uint8_t *)q + shmHeaderLen + curHead + padding);
    uint32_t itemLen = ATOMIC_LOAD(nextItemPtr);
    if (itemLen == 0) {
      if (curTail > curHead) {
        ATOMIC_CAS(&(q->Consumer), &curHead, curTail);
      } else {
        ATOMIC_CAS(&(q->Consumer), &curHead, 0);
      }
      continue;
    }
    *newHead = curHead + padding + itemHeaderLen + (uint64_t)(itemLen);
    if (curTail < curHead) {
      if (curHead + minLeftLen + (uint64_t)(itemLen) > Len(q)) {
        return ErrOfHeadOutOfRange;
      }
    } else {
      if (*newHead > curTail) {
        return ErrOfHeadOutOfRange;
      }
    }
    *outLen = itemLen;
    *dst = (uint8_t *)q + shmHeaderLen + curHead + padding + itemHeaderLen;
    return GetOneSuccess;
  }
}

bool CommitConsume(SpscQueue *q, uint64_t newHead) {
  uint64_t curHead = ATOMIC_LOAD(&(q->Consumer));
  return ATOMIC_CAS(&(q->Consumer), &curHead, newHead);
}

int Consume(SpscQueue *q, void *dst, uint32_t *len) {
  void *buf = NULL;
  uint32_t itemLen = 0;
  uint64_t newHead = 0;
  int ret = GetOne(q, &buf, &itemLen, &newHead);
  if (ret != GetOneSuccess) {
    return ret;
  }
  if (*len < itemLen) {
    return ErrOfOutBufferTooSmall;
  }
  *len = itemLen;
  memcpy(dst, buf, itemLen);
  if (!CommitConsume(q, newHead)) {
    return ErrOfCommitConsumeFail;
  }
  return GetOneSuccess;
}

void Unmap(SpscQueue *q) { munmap(q, Len(q) + shmHeaderLen); }
